[SPARK-18929][ML] Add Tweedie distribution in GLM

actuaryzhang opened the pull request:
## What changes were proposed in this pull request?
I propose to add the full Tweedie family into the GeneralizedLinearRegression model. The Tweedie family is characterized by a power variance function. Currently supported distributions such as Gaussian, Poisson and Gamma families are a special case of the Tweedie https://en.wikipedia.org/wiki/Tweedie_distribution.

@yanboliang @srowen @sethah

actuaryzhang committed:
Add Tweedie family to GLM

actuaryzhang committed:
Fix calculation in dev resid; Add test for different var power

actuaryzhang committed:
Merge test into GLR

AmplabJenkins commented:
Can one of the admins verify this patch?

actuaryzhang commented:
Jenkins, add to whitelist

yanboliang commented:
Jenkins, add to whitelist

yanboliang commented:
ok to test

yanboliang commented:
Jenkins, test this please.

srowen commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -592,6 +629,59 @@ object GeneralizedLinearRegression extends DefaultParamsReadable[GeneralizedLine
+      if (variancePower > 1.0 && variancePower < 2.0) {
+        require(y >= 0.0, "The response variable of the specified Tweedie distribution " +
+          s"should be non-negative, but got $y")
+        math.max(y, 0.1)
```
If we're going to use this magic 0.1 constant in many places, factor out a constant? 0.1 seems quite large as an 'epsilon' but I guess that's what R's implementation uses for whatever reason?

actuaryzhang replied:
I have not seen a formal justification for the choice of 0.1 in R. This seminal [paper](http://users.du.se/~lrn/StatMod10/HomeExercise2/Nelder_Pregibon.pdf) suggests 1/6 (about 0.17) to be the best constant. I would prefer to be consistent with R so that we can make comparison. Using a constant is a good idea.

srowen commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -592,6 +629,59 @@ object GeneralizedLinearRegression extends DefaultParamsReadable[GeneralizedLine
+        deviance: Double,
+        numInstances: Double,
+        weightSum: Double): Double = {
+      0.0
```
Throw a UnsupportedOperationException?

srowen commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -592,6 +629,59 @@ object GeneralizedLinearRegression extends DefaultParamsReadable[GeneralizedLine
+      if (mu < epsilon) {
+        epsilon
+      } else if (mu.isInfinity) {
+        Double.MaxValue
```
Out of curiosity is this meaningful to "cap" at Double.MaxValue? By the time you get there a lot of stuff is going to be infinite or not meaningful.

srowen replied:
I see, it's done that way in other implementations. OK. I'm not sure if it's going to do much.

I think there's a problem in the Gaussian project method because it uses Double.MinValue to appear to mean "the smallest double" when it's the "smallest possible double" I'll investigate and file a bug if needed.

srowen commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -592,6 +629,59 @@ object GeneralizedLinearRegression extends DefaultParamsReadable[GeneralizedLine
+
+    val defaultLink: Link = Log
+
+    var variancePower: Double = 1.5
```
This is a global shared variable -- we really can't do this.

actuaryzhang replied:
Would you please suggest a better way to set the variancePower? I want to be consistent with the existing code to have the `Family` objects, but I need to also pass on the input `variancePower` to the `Tweedie` object which is used to compute the variance function. Any suggestion will be highly appreciated. @srowen @yanboliang

srowen replied:
I think the Tweedie implementation needs to be able to access parameters of the GLM, to read off variancePower.

As it is this is a global variable and two jobs would overwrite each others' values.

srowen reviewed (changes requested):
That's probably reasonable. Doesn't this give us support for compound Poisson et al, by supporting general Tweedie distributions?

srowen commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -242,7 +275,7 @@ class GeneralizedLinearRegression @Since("2.0.0") (@Since("2.0.0") override val
 
   override protected def train(dataset: Dataset[_]): GeneralizedLinearRegressionModel = {
-    val familyObj = Family.fromName($(family))
+    val familyObj = Family.fromName($(family), $(variancePower))
```
I don't think we can do this either. variancePower is specific to one family, not a property of all of them.

actuaryzhang committed:
Use Tweedie class instead of global object Tweedie; change variancePower to varPower

actuaryzhang commented:
@srowen Thanks for the comments. Really helpful. I have made a new commit that addresses the issues you raised:

- I think the use of a global family object does not work well for the tweedie case, since we have to set the variance power. I now define a tweedie class and create the tweedie object within the `train` method of `GeneralizedLinearRegression` where the variance power is set. Does this make sense? 
- I created a constant `delta = 0.1` in the `Family` class which is used to shift `mu (or y) = 0` to avoid numerical issues. 
- I now change `variancePower` to `varPower` to save typing (and consistent with R and H2o).
- The `project` method you asked about is copied from the other families, which I suppose is to bound the mean to stabilize estimation.
- I now throw `UnsupportedOperationException` for tweedie AIC.

Let me know if there is any other issues. 

Copying others for reviewing. @yanboliang @sethah

yanboliang commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -303,14 +341,15 @@ object GeneralizedLinearRegression extends DefaultParamsReadable[GeneralizedLine
-    Binomial -> Logit, Binomial -> Probit, Binomial -> CLogLog,
-    Poisson -> Log, Poisson -> Identity, Poisson -> Sqrt,
-    Gamma -> Inverse, Gamma -> Identity, Gamma -> Log
+    "gaussian" -> Identity, "gaussian" -> Log, "gaussian" -> Inverse,
```
String is error-prone, I think we can construct a member object for ```Tweedie``` whose ```variancePower``` is the default value(1.5).

actuaryzhang replied:
@yanboliang  The member object (in `GeneralizedLinearRegressionBase`) won't be accessible in `Family`, right? The method `Family.fromName($(family))` uses global objects like `Poisson`, `Gamma` etc. To use  `Family.fromName`, I need to create a `Tweedie` global object. Then we are back to the issue that @srowen pointed out of setting `variancePower` of the global object. Please advise. Thanks.

actuaryzhang replied:
@yanboliang Could you help me understand the issue caused by using string? If I use object, then I have to create a Tweedie object that is not used anywhere else. And also I have to write two methods in `Family`: one returns the global Tweedie object (where the variancePower is preset) and one returns the a TweedieFamily object created using the user-specified variancePower. I hope we are fine using string since there are only a few values.

yanboliang replied:
Yeah, you are right. It needs an extra global object to avoid error-prone which may a little expensive. I'm ok with using string.

actuaryzhang commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -242,7 +275,12 @@ class GeneralizedLinearRegression @Since("2.0.0") (@Since("2.0.0") override val
   override protected def train(dataset: Dataset[_]): GeneralizedLinearRegressionModel = {
-    val familyObj = Family.fromName($(family))
+    val familyObj = if ($(family) == "tweedie") {
+      new Tweedie($(varPower))
```
`Family` is not subclass of `GeneralizedLinearRegression`. Could you elaborate how to make it get the `variancePower` value?

actuaryzhang committed:
Allow Family to use GLRBase object directly

actuaryzhang committed:
Add TweedieFamily and implement specific distn within Tweedie

actuaryzhang committed:
Clean up doc

actuaryzhang commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -64,6 +64,27 @@ private[regression] trait GeneralizedLinearRegressionBase extends PredictorParam
   def getFamily: String = $(family)
 
   /**
+   * Param for the power in the variance function of the Tweedie distribution which provides
```
changed tweedie. but other docs have been using Param..

actuaryzhang commented:
@srowen @yanboliang Thanks much for the feedback. I now have a better understanding of the code and the issue. I have made new commits reflecting your suggestions. The major changes are summarized below:

- Create new method `fromModel` in the `Family` object to get the desired family object by accessing the `GeneralizedLinearRegressionBase` object directly. This solves the main challenge of setting the variancePower of the Tweedie family. The old `fromName` method is deleted since we now need both `name` and `variancePower`. 
- Create `TweedieFamily` as new subclass of `family`. This class encompasses the special cases of Gaussian, Poisson and Gamma, which can be handled together using the general formula based on variancePower. 

Please let me know if there are any other issues. Thanks.

srowen commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -303,20 +337,24 @@ object GeneralizedLinearRegression extends DefaultParamsReadable[GeneralizedLine
   private[regression] val epsilon: Double = 1E-16
 
+  /** Constant used in initialization and deviance to avoid numerical issues. */
+  private[regression] val delta: Double = 0.1
```
This should still be in an `object` IMHO; it's a constant right? `epsilon` really should be too. It's not a big deal but not quite right.

actuaryzhang replied:
They are already in the `GeneralizedLinearRegression` object, aren't they? Or do you mean creating a new object say `Constant` that stores these two constants, and using them like `Constant.delta`? 

Since `delta` is only used in the `TweedieFamily` class, I can also move it there. Let me know what is best. Thanks.

yanboliang replied:
I'm OK to put it here.

srowen replied:
Why not a companion object for TweedieFamily though? that just seems easy and more correct

srowen commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -397,49 +436,132 @@ object GeneralizedLinearRegression extends DefaultParamsReadable[GeneralizedLine
+    }
 
-    override def initialize(y: Double, weight: Double): Double = y
+    override def variance(mu: Double): Double = {
```
Instead of case statements like this, why not just override in subclasses?

srowen commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -397,49 +436,132 @@ object GeneralizedLinearRegression extends DefaultParamsReadable[GeneralizedLine
-        weightSum: Double): Double = {
-      val wt = predictions.map(x => math.log(x._3)).sum()
-      numInstances * (math.log(deviance / numInstances * 2.0 * math.Pi) + 1.0) + 2.0 - wt
+    override def aic(predictions: RDD[(Double, Double, Double)],
```
Likewise there's not a lot of value in pushing 4 separate implementations into one method and using if-else. Just override this method.

srowen commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -397,49 +436,132 @@ object GeneralizedLinearRegression extends DefaultParamsReadable[GeneralizedLine
+  private[regression] class TweedieFamily(private val variancePower: Double)
+    extends Family{
+
+    val name: String = variancePower match {
```
Why remove the name and switch like this? you can instead adjust this so that subclasses override a `name` method.

srowen reviewed (changes requested):
I think this is moving somewhere interesting, though I'm still not sure about this design. Comments inline.

actuaryzhang committed:
Move defaultLink and name to subclass of TweedieFamily

actuaryzhang committed:
Change style for AIC

actuaryzhang commented:
@srowen Thanks for the comments. Makes lots of sense to move the switch to subclass. I did not know one could override a `val`. 

In the new commit, I have moved the `defaultLink` and `aic` to subclasses. I also added back `name` in the constructor. The other methods `initialize`, `variance`, `deviance` and `project` are common and thus implemented in the `TweedieFamily` parent class. Hope this makes sense.

yanboliang commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -397,46 +434,118 @@ object GeneralizedLinearRegression extends DefaultParamsReadable[GeneralizedLine
-        case Binomial.name => Binomial
-        case Poisson.name => Poisson
-        case Gamma.name => Gamma
+    def fromModel(model: GeneralizedLinearRegressionBase): Family = {
```
Rename to ```fromParams```, we extract ```family``` and ```variancePower``` from the ```Params``` which is the superclass of GLR estimator and model. And actually we use this function for both estimator(L279) and model(L974).

yanboliang commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -397,46 +434,118 @@ object GeneralizedLinearRegression extends DefaultParamsReadable[GeneralizedLine
+    * Tweedie exponential family distribution.
+    * This includes the special cases of Gaussian, Poisson and Gamma.
+    */
+  private[regression] class TweedieFamily(private val variancePower: Double)
```
```TweedieFamily``` -> ```Tweedie```, we don't add suffix for other family class/object.

yanboliang commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -397,46 +434,118 @@ object GeneralizedLinearRegression extends DefaultParamsReadable[GeneralizedLine
+    extends Family("tweedie") {
+
+    /*
+      The canonical link is 1 - variancePower. Except for the special cases of Gaussian,
```
```The canonical link is 1 - variancePower```, could you clarify this to make it more clear?

yanboliang commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -397,46 +434,118 @@ object GeneralizedLinearRegression extends DefaultParamsReadable[GeneralizedLine
-    val defaultLink: Link = Identity
+    override def initialize(y: Double, weight: Double): Double = {
+      if (variancePower >= 1.0 && variancePower < 2.0) {
+        require(y >= 0.0, s"The response variable of the specified distribution " +
```
```The response variable of $name family ```

yanboliang commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -397,46 +434,118 @@ object GeneralizedLinearRegression extends DefaultParamsReadable[GeneralizedLine
+        require(y >= 0.0, s"The response variable of the specified distribution " +
+          s"should be non-negative, but got $y")
+      } else if (variancePower >= 2.0) {
+        require(y > 0.0, s"The response variable of the specified distribution " +
```
Ditto.

yanboliang commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -397,46 +434,118 @@ object GeneralizedLinearRegression extends DefaultParamsReadable[GeneralizedLine
+   * Gaussian exponential family distribution.
+   * The default link for the Gaussian family is the identity link.
+   */
+  private[regression] object Gaussian extends TweedieFamily(0.0) {
```
We should keep the concrete implementation of ```variance, deviance and aic``` for Gaussian, Poisson and Gamma, the main reasons are:
* These functions were called very frequently, the concrete implementation in subclasses should be more efficient.
* It's helpful to locate errors or bugs.

yanboliang commented on mllib/src/test/scala/org/apache/spark/ml/regression/GeneralizedLinearRegressionSuite.scala:
```diff
@@ -578,6 +578,100 @@ class GeneralizedLinearRegressionSuite
 
+  test("generalized linear regression: tweedie family against glm") {
+    /*
+        R code:
```
```library(statmod)``` which can help users to reproduce this test case.

yanboliang reviewed (commented):
@actuaryzhang Please see my inline comments. Thanks.

actuaryzhang committed:
Rename Family methods and restore methods for tweedie subclasses

actuaryzhang committed:
Update test

actuaryzhang commented:
@yanboliang Thanks much for the detailed comments. I have addressed all of them in the new commits. Please take another look. 
@srowen

actuaryzhang commented:
@srowen @yanboliang 
Any additional issues regarding this PR?

actuaryzhang committed:
Clean up doc

srowen commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -397,21 +435,102 @@ object GeneralizedLinearRegression extends DefaultParamsReadable[GeneralizedLine
+     * 2) if family name is tweedie, retrieve object based on variancePower
      *
-     * @param name family name: "gaussian", "binomial", "poisson" or "gamma".
+     * @param params a GenerealizedLinearRegressionBase object
```
typo in GenerealizedLinearRegressionBase; this type of doc doesn't do anything though because the type is already documented. It should say something non-trivial

srowen reviewed (commented):
This is down to small items now, I think it's pretty good.

actuaryzhang committed:
Put delta in Tweedie companion object

actuaryzhang commented:
@srowen Made a new commit according to your suggestion. Everything looking good now? 
@yanboliang

actuaryzhang committed:
Clean up doc

yanboliang commented:
@actuaryzhang I'm on travel these days, and will make another pass in a few days. Thanks.

actuaryzhang commented:
@yanboliang Did you get a chance to take another look at this? Thanks.

yanboliang commented:
@actuaryzhang I just came back from vacation, and will take a look in two days. Thanks.

actuaryzhang committed:
Allow more link functions in tweedie

actuaryzhang commented:
@yanboliang Thanks for the detailed review. I have made all changes you suggested except for the part on the new power link function. Yes, the canonical link in the Tweedie in general is `1.0 - variancePower`. But other than the special cases of 'identity', 'log', 'inverse', it is rarely used. The implementation of the generic power link function is unnecessary in my opinion for the following reasons:

1. The default link in the Tweedie is overridden in Gaussian, Poisson and Gamma to the respective canonical link. So when one specifies tweedie with variancePower = 0, 1, 2, they will have the correct canonical link.  I now allow Tweedie to take link function 'identity', 'log', 'inverse', 'sqrt', which cover almost all possible links that will be used in practice.  
2. If we allow an additional 'linkPower" parameter, we will have two ways to specify the link: one through 'link' and one through 'linkPower'. This will be confusing. 
3. The implementation of the 'linkPower' is messy with almost no additional gain. Have to write new powerLink class, fromParams method which determines the link function based on 'link', 'linkPower' as well as 'family', and also various checks (consistency between link and linkPower, supported Family and link, etc). I tried it, and it's very ugly. 
 
If we really want a power link function, can I suggest that we implement it in a separate PR since it is really not specific to the 'tweedie' family? It can also be used together with other distributions. 

@srowen

actuaryzhang commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -158,6 +183,16 @@ class GeneralizedLinearRegression @Since("2.0.0") (@Since("2.0.0") override val
+    */
+  @Since("2.2.0")
+  def setVariancePower(value: Double): this.type = set(variancePower, value)
+  setDefault(variancePower -> 1.5)
```
Done. change default variancePower to 0.0, which will use Gaussian (with default identity link)

actuaryzhang commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -365,7 +401,6 @@ object GeneralizedLinearRegression extends DefaultParamsReadable[GeneralizedLine
   /**
    * A description of the error distribution to be used in the model.
    *
-   * @param name the name of the family.
```
Sorry. added this back.

actuaryzhang commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -397,32 +432,121 @@ object GeneralizedLinearRegression extends DefaultParamsReadable[GeneralizedLine
+            case 0.0 => Gaussian
+            case 1.0 => Poisson
+            case 2.0 => Gamma
+            case default => new Tweedie(default)
```
Done.

yanboliang commented:
@actuaryzhang I totally agree that other link functions are rarely used for ```tweedie``` family, but it should be supported correctly for all possible input. I list the reasons why we should expose ```linkPower``` param:
*
```
val trainer = new GeneralizedLinearRegression().setFamily("tweedie").setVariancePower(1.5)
```
should produce the same output compared with
```
glm(formula = "b ~ .", family = tweedie(var.power=1.5), data = df)
```
unfortunately they are different in the current implementation, since the default link function for ```tweedie``` family with ```variancePower=1.5``` is ```log``` in your code, which is incorrect, it should be ```mu^-0.5```.
* I think the driving force of this PR is to support other distributions and links except the existing ones. If users want to use Gaussian, Poisson and Gamma families, they are likely to specify corresponding names directly rather than using ```tweedie``` family with ```variancePower = 0, 1, 2```. So if users specify family with ```tweedie```, they are likely to use other distributions and link functions which we should support correctly.
* To your second point, if we allow an additional ```linkPower``` parameter, we can add documents to ```link``` to clarify it only applicable when family is not ```tweddie```. If family was set as ```tweddie```, we use ```linkPower``` to choose the link function.

Please feel free to let me know your thoughts, thanks.

actuaryzhang commented:
@yanboliang Thanks for the feedback. However, I'm not sure why we need to be consistent with R on this one. The usage of 'tweedie' glm almost always uses `link.power = 0, 1, -1`. There is barely any use case for the default `1-variancePower` link.  I think a better design is to set the default to the one that is used more often, and make it clear that `we only support identity, log, inverse and sqrt link functions`. 

On the other hand, even if we want to be consistent with R, I actually wrote the R package `cplm` which fits the tweedie [glm](http://finzi.psych.upenn.edu/library/cplm/html/cpglm.html) as well as [mixed effect model](https://www.rdocumentation.org/packages/cplm/versions/0.7-4/topics/cpglmm). In that package I set the default to the `log` link and never had an issue. 

I don't think I will implement `linkPower`, but I'm fine if you want to do it. However, if the lack of `linkPower` means this PR won't get merged, please let me know and we can close this and move on.

actuaryzhang commented:
@srowen @yanboliang 
I'm closing this PR since it does not seem to be very clean to integrate into the current GLM setup. I appreciate all the comments and discussions.

actuaryzhang committed:
Implement link power

AmplabJenkins commented:
Can one of the admins verify this patch?

actuaryzhang commented:
Jenkins, test this please

actuaryzhang commented:
Sorry about closing this prematurely. I'm giving it another shot and I think I have an elegant solution to include `linkPower`. The new commit adds the following:
1. It implements the broad family of power link function, specified through `linkPower`. There is now a `PowerLink` class for the power link function. It has subclasses `Identity`, `Log`, `Inverse` and `Sqrt`. With this, the GLM now supports all distributions characterized by power variance function and power link function. For now, I restrict the `linkPower` to be in `[-10, 10]` for numerical stability, but can change that.  
2. The key to avoid all the messy coding is to use **only** `link` for non-tweedie family, and **only** `lnkPower` for tweedie family, as @yanboliang suggested. I have added validation for this, and also the `fromParams` method in `Link` to get the correct Link object based on `link` and `linkPower`. 
3. I added new tests to test tweedie with default link. For example, this reproduces the Gaussian GLM estimate when `variancePower = 0` (default link would be identity). Similarly, this addresses @yanboliang example where the result from `val trainer = new GeneralizedLinearRegression().setFamily("tweedie").setVariancePower(1.5)` produces the same estimate as in R `glm(formula = "b ~ .", family = tweedie(var.power=1.5), data = df)`. 

@yanboliang @srowen Would you please take another look and let me know if there is additional changes needed? Thanks.

actuaryzhang committed:
remove restriction on link power; revert to use family object in check

yanboliang commented:
@actuaryzhang Thanks the update, I will have a look before this weekend.

actuaryzhang commented:
@yanboliang Thanks. Look forward to your feedback.

yanboliang commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -106,11 +148,20 @@ private[regression] trait GeneralizedLinearRegressionBase extends PredictorParam
-        s"with ${$(family)} family does not support ${$(link)} link function.")
+    if ($(family) == "tweedie") {
+      require(!isDefined(link), "The link function for the tweedie family must be " +
+        "specified using linkPower, not link.")
```
I don't think we should throw error if users set ```link``` when family is set as ```tweedie```, and a warning log should be okay, like
```
if (isSet(link)) {
        logWarning("When family is tweedie, use param linkPower to specify link function. " +
          "Setting param link will take no effect.")
      }
```

yanboliang commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -106,11 +148,20 @@ private[regression] trait GeneralizedLinearRegressionBase extends PredictorParam
+        "specified using linkPower, not link.")
+    } else {
+      require(!isDefined(linkPower), s"The link function for the ${$(family)} family " +
+      "must be specified using link, not linkPower.")
```
Ditto,
```
if (isSet(linkPower)) {
        logWarning("When family is not tweedie, use param link to specify link function. " +
          "Setting param linkPower will take no effect.")
      }
```

yanboliang commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -613,25 +758,67 @@ object GeneralizedLinearRegression extends DefaultParamsReadable[GeneralizedLine
 
     /**
-     * Gets the [[Link]] object from its name.
+     * Gets the [[Link]] object based on link and linkPower.
```
and -> or

yanboliang commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -613,25 +758,67 @@ object GeneralizedLinearRegression extends DefaultParamsReadable[GeneralizedLine
+  }
+
+  /** Power link function class */
+  private[regression] class PowerLink(val linkPower: Double)
```
```PowerLink -> Power```, other link objects are not end with ```Link```.

yanboliang commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -242,9 +316,9 @@ class GeneralizedLinearRegression @Since("2.0.0") (@Since("2.0.0") override val
+      Link.fromParams(this)
     } else {
       familyObj.defaultLink
     }
```
This code snippets are used multiple places, could we wrapper them up as a function?
```
def getFamilyAndLinkObj(params: GeneralizedLinearRegressionBase): (Family, Link) = {
......
}
```

actuaryzhang replied:
Makes sense. I created a companion object `FamilyAndLink` with a factory method to construct desired `FamilyAndLink` objects from the input param map.

yanboliang commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -613,25 +758,67 @@ object GeneralizedLinearRegression extends DefaultParamsReadable[GeneralizedLine
+          case -1.0 => Inverse
+          case 0.5 => Sqrt
+          case others => new PowerLink(others)
+        }
```
There is no default value for ```linkPower```, what happened if users don't set it?

actuaryzhang replied:
The last version of the code only allows setting `linkPower` for tweedie, or `link` for non-tweedie. But now since we only give warnings, I need to add the logic for checking the correct specification. In the `FamilyAndLink` companion object, I use the following logic. When family is not tweedie and link is set,  or when family is tweedie and linkPower is set, I will use `Link.fromParams` to construct the Link object. Otherwise, use the default link from the corresponding family.  

So, if the user does not specify `linkPower`,  the link object will be the default `new Power(1 - variancePower)` set in the `Tweedie` class. The test for `tweedie family against glm (default power link)` covers this. The reason for not setting a default linkPower in the param map is to mimic the existing behavior: for tweedie with variancePower = 0, this will be the Gaussian with the identity link;  for tweedie with variancePower = 1, this will be the Poisson with the log link; etc.   

```
     val linkObj = if ((params.getFamily != "tweedie" && params.isDefined(params.link)) ||
        (params.getFamily == "tweedie" && params.isDefined(params.linkPower))) {
        Link.fromParams(params)
      } else {
        familyObj.defaultLink
      }
```

yanboliang commented:
@actuaryzhang The overall architecture looks good now and I left some comments. I did not review the tests carefully, since I think we should trigger Jenkins to run the tests firstly. Thanks.

yanboliang commented:
Jenkins, test this please.

actuaryzhang committed:
create factory method for FamilyAndLink

actuaryzhang commented:
@yanboliang Thanks for the review and comments. I have made a new commit that addressed all your comments. The main change is the new companion object `FamilyAndLink` and factory methods to construct desired `FamilyAndLink` objects from input param map.  Please see my inline responses for details.

SparkQA commented:
**[Test build #3531 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3531/testReport)** for PR 16344 at commit [`6f4abeb`](https://github.com/apache/spark/commit/6f4abebb7a327b508fdc48789793ca6a29dc9842).

SparkQA commented:
**[Test build #3531 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3531/testReport)** for PR 16344 at commit [`6f4abeb`](https://github.com/apache/spark/commit/6f4abebb7a327b508fdc48789793ca6a29dc9842).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

actuaryzhang committed:
fix style issue in test

actuaryzhang commented:
Jenkins, test this please.

actuaryzhang commented:
Jenkins, retest this please

yanboliang commented:
ok to test.

yanboliang commented:
add to whitelist.

yanboliang commented:
Jenkins, test this please.

yanboliang commented:
Jenkins, test this please.

yanboliang commented:
It's weird that I can't trigger this job as well. Ping @srowen @jkbradley @MLnick, could you add @actuaryzhang to whitelist to trigger this job. Thanks.

actuaryzhang committed:
make model writable

felixcheung commented:
Jenkins, retest this please

yanboliang commented:
I can't figure out why I couldn't add you to whitelist and trigger this job. Ping @rxin @srowen Could you help to add @actuaryzhang to whitelist to trigger this job? Thanks.

actuaryzhang commented:
Still not testing... Been in the status "Asked to test" for a few days now. 
How can we resolve this? Please help kick off the test. Thanks!
@yanboliang @felixcheung @srowen @rxin @jkbradley

SparkQA commented:
**[Test build #3536 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3536/testReport)** for PR 16344 at commit [`5a40073`](https://github.com/apache/spark/commit/5a4007330c4c8187a00d6d31a9d8a3dcb1798fc9).

SparkQA commented:
**[Test build #3536 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3536/testReport)** for PR 16344 at commit [`5a40073`](https://github.com/apache/spark/commit/5a4007330c4c8187a00d6d31a9d8a3dcb1798fc9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

actuaryzhang committed:
Merge branch 'master' into tweedie

actuaryzhang committed:
resolve conflicts

actuaryzhang commented:
Jenkins, retest this please

actuaryzhang commented:
@yanboliang Finally, the test is done. Is there anything else needed for this PR?

actuaryzhang commented:
@srowen @yanboliang @felixcheung Could you help kick off the new test please? Seems to be hanging for a day now. Thanks much.

actuaryzhang commented:
@srowen @yanboliang @felixcheung @jkbradley Could you help kick off the new test please? Thanks.

MLnick commented:
jenkins add to whitelist

MLnick commented:
jenkins test this please

SparkQA commented:
**[Test build #71643 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71643/testReport)** for PR 16344 at commit [`83deee3`](https://github.com/apache/spark/commit/83deee352c46ec113554fccee4bdc14ead56072e).

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71643/
Test FAILed.

actuaryzhang commented:
Could anybody help me understand what's causing this test to fail? I see several other ML PR failing as well, with the same error message like below:

> Error instrumenting class:org.apache.spark.deploy.mesos.ui.MesosClusterUI
Error instrumenting class:org.apache.spark.mllib.regression.IsotonicRegressionModel$SaveLoadV1_0$
> 

Was this related to the recent commit on instrumentation #15671? How do I fix this? Thanks! 
@zhengruifeng @yanboliang

yanboliang commented:
Jenkins, retest this please.

SparkQA commented:
**[Test build #71686 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71686/testReport)** for PR 16344 at commit [`83deee3`](https://github.com/apache/spark/commit/83deee352c46ec113554fccee4bdc14ead56072e).

yanboliang commented:
@actuaryzhang This test failure is caused by Jenkins was not stable, you just need to retest if you encounter similar issue.

SparkQA commented:
**[Test build #71686 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71686/testReport)** for PR 16344 at commit [`83deee3`](https://github.com/apache/spark/commit/83deee352c46ec113554fccee4bdc14ead56072e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71686/
Test PASSed.

yanboliang commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -409,27 +503,108 @@ object GeneralizedLinearRegression extends DefaultParamsReadable[GeneralizedLine
-     * Gets the [[Family]] object from its name.
+     * Gets the [[Family]] object based on family and variancePower.
+     * 1) retrieve object based on family name
+     * 2) if family name is tweedie, retrieve object based on variancePower
```
Should the following document be better?
```
Gets the Family object based on param family and variancePower.
If param family was set with "gaussian", "binomial", "poisson" or "gamma", return the corresponding object directly; otherwise, construct a Tweedie object according to variancePower.
```

yanboliang commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -620,25 +779,67 @@ object GeneralizedLinearRegression extends DefaultParamsReadable[GeneralizedLine
 
     /**
-     * Gets the [[Link]] object from its name.
+     * Gets the [[Link]] object based on link or linkPower.
```
```
Gets the Link object based on param family, link and linkPower.
If param family was set with "tweedie", return or construct link function object according to linkPower; otherwise, return link function object according to link.
```

yanboliang commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -620,25 +779,67 @@ object GeneralizedLinearRegression extends DefaultParamsReadable[GeneralizedLine
      *
-     * @param name link name: "identity", "logit", "log",
-     *             "inverse", "probit", "cloglog" or "sqrt".
+     * @param params the parameter map containing link and link power
```
```link and link power``` -> ```family, link and linkPower```

yanboliang commented on mllib/src/test/scala/org/apache/spark/ml/regression/GeneralizedLinearRegressionSuite.scala:
```diff
@@ -1052,6 +1217,120 @@ class GeneralizedLinearRegressionSuite
+        1.0, 3.0, 2.0, 1.0,
+        0.0, 4.0, 3.0, 3.0), 4, 4, byrow = TRUE))
+
+      f <- glm(V1 ~ -1 + V3 + V4, data = df, weights = V2,
```
Change ```f``` to ```model```, and add ```summary(model)``` at next line. It's helpful for users to reproduce the result in R.

yanboliang commented on mllib/src/test/scala/org/apache/spark/ml/regression/GeneralizedLinearRegressionSuite.scala:
```diff
@@ -1052,6 +1217,120 @@ class GeneralizedLinearRegressionSuite
+
+      Number of Fisher Scoring iterations: 11
+
+      residuals(model, type="pearson")
```
Typos here? I guess you paste irrelevant results for the following ```residuals```, they should be consistent with L1279-L1281.

yanboliang reviewed (commented):
Only some comments about documents, otherwise, looks good to me. Thanks!

actuaryzhang committed:
Merge branch 'master' into tweedie

SparkQA commented:
**[Test build #71817 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71817/testReport)** for PR 16344 at commit [`995e88f`](https://github.com/apache/spark/commit/995e88ffcc5584e468fd7fb356bb3aab31ec0217).

actuaryzhang committed:
update docs

SparkQA commented:
**[Test build #71817 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71817/testReport)** for PR 16344 at commit [`995e88f`](https://github.com/apache/spark/commit/995e88ffcc5584e468fd7fb356bb3aab31ec0217).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71817/
Test PASSed.

SparkQA commented:
**[Test build #71823 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71823/testReport)** for PR 16344 at commit [`54da2cb`](https://github.com/apache/spark/commit/54da2cbbb53ddde3a91ef6d0d98128d8c7f3deb8).

actuaryzhang commented:
@yanboliang Thanks so much for your detailed review. Your suggestions make lots of sense and I have included all of them in the new commit. Let me know if there is any other change needed.

SparkQA commented:
**[Test build #71823 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71823/testReport)** for PR 16344 at commit [`54da2cb`](https://github.com/apache/spark/commit/54da2cbbb53ddde3a91ef6d0d98128d8c7f3deb8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71823/
Test PASSed.

yanboliang commented:
This looks good to me. cc @jkbradley @srowen @sethah for another pass if you have time. Thanks.

yanboliang commented:
Merged into master. If there are comments from others, we can address them in follow-up work. Thanks.
